flamedisx/likelihood.py

```diff
@@ -263,64 +310,74 @@ def _param_i(self, pname):
+                i_batch, params, dsetname, data_tensor, batch_info,
+                autograph=True)
         grad = t.gradient(ll, grad_par_list)
         return ll, tf.stack(grad)
 
-    def _log_likelihood_grad2(self, i_batch, dsetname, autograph,
+    def _log_likelihood_grad2(self, i_batch, dsetname, data_tensor, batch_info,
                               omit_grads=tuple(), **params):
-        if omit_grads is None:
-            omit_grads = []
         grad_par_list = [x for k, x in params.items()
                          if k not in omit_grads]
         with tf.GradientTape(persistent=True) as t2:
             t2.watch(grad_par_list)
             with tf.GradientTape() as t:
                 t.watch(grad_par_list)
-                ll = self._log_likelihood_inner(i_batch, params,
-                                                dsetname, autograph)
+                ll = self._log_likelihood_inner(
+                    i_batch, params, dsetname, data_tensor, batch_info,
+                    autograph=False)
             grads = t.gradient(ll, grad_par_list)
         hessian = [t2.gradient(grad, grad_par_list) for grad in grads]
         del t2
         return ll, tf.stack(grads), tf.stack(hessian)
 
-    def _log_likelihood_inner(self, i_batch, params, dsetname, autograph):
-        # Does for loop over datasets and sources, not batches
-        # Sum over sources is first in likelihood
+    def _log_likelihood_inner(self, i_batch, params,
+                              dsetname, data_tensor, batch_info,
+                              autograph):
+        """Return log likelihood contribution of one batch in a dataset
+
+        This loops over sources in the dataset and events in the batch,
+        but not not over datasets or batches.
+        """
+        # Retrieve batching info. Cannot use tuple-unpacking, tensorflow
+        # doesn't like it when you iterate over tenstors
+        dataset_index = self.dsetnames.index(dsetname)
+        n_batches = batch_info[dataset_index, 0]
+        batch_size = batch_info[dataset_index, 1]
+        n_padding = batch_info[dataset_index, 2]
 
         # Compute differential rates from all sources
         # drs = list[n_sources] of [n_events] tensors
-        drs = tf.zeros((self.batch_size[dsetname],), dtype=fd.float_type())
-        for sname, s in self.sources.items():
-            if not self.d_for_s[sname] == dsetname:
-                continue
+        drs = tf.zeros((batch_size,), dtype=fd.float_type())
+        for source_i, sname in enumerate(self.sources_in_dset[dsetname]):
+            s = self.sources[sname]
             rate_mult = self._get_rate_mult(sname, params)
-            dr = s.differential_rate(s.data_tensor[i_batch],
-                                     autograph=autograph,
-                                     **self._filter_source_kwargs(params, sname))
+
+            col_start, col_stop = self.column_indices[dsetname][source_i]
```

Comment:
`column_indices` is constant for the lifetime of the Source object right?

Reply:
Yes, sources can't change their mind about which columns they use